### DISCUSSION TITLE
oidc not able to login with cli with KeyCloak

### DISCUSSION THREAD

**Discussion by gnometsunami** [argoproj/argo-cd#6993] [Q&A] [answered]
Is argocd cli not supported with oidc? Or do I have something misconfigured?

UI login works fine, but cli login gives this error:
```
$ argocd login argocd.my.domain --grpc-web --sso
Opening browser for authentication
Performing authorization_code flow login: <cut for brevity>
FATA[0001] oauth2: cannot fetch token: 401 Unauthorized
Response: {"error":"unauthorized_client","error_description":"Client secret not provided in request"}
```
server logs:
```
time="2021-08-14T22:35:21Z" level=info msg="received unary call /cluster.SettingsService/Get" grpc.method=Get grpc.request.claims=null grpc.request.content= grpc.service=cluster.SettingsService grpc.start_time="2021-08-14T22:35:21Z" span.kind=server system=grpc
time="2021-08-14T22:35:21Z" level=info msg="Ignore status for CustomResourceDefinitions"
time="2021-08-14T22:35:21Z" level=info msg="finished unary call with code OK" grpc.code=OK grpc.method=Get grpc.service=cluster.SettingsService grpc.start_time="2021-08-14T22:35:21Z" grpc.time_ms=1.78 span.kind=server system=grpc
```

**Comment by jessesuen** [comment]
Argo CD should support OIDC CLI login. Could you provide more details about your setup? who is your provider? are you using dex or native oidc support? Are you able to paste the config (redacted)?

**Reply by gnometsunami** [reply]
I am using keycloak as my idp. I'm pretty much blindly following the Argo documentation. Argo was installed and keycloak signin configured as per the docs. The only deviation Ive made was to configure an ingress for tls termination instead of a loadbalancer service.

I think I'm using dex? How can I tell? I'm not totally sure what dex is and how it fits into the application.

**Reply by jessesuen** [reply]
Regarding Dex, we use Dex as an SSO **_option_**, to allow users to delegate things like GitHub, SAML, LDAP as an OIDC provider (despite those things not traditionally considered an OIDC providers). Dex allows translation of things like GitHub teams into OIDC groups. You can read more here:
https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/#dex

In any case, if you are following the [keycloak docs](https://argo-cd.readthedocs.io/en/stable/operator-manual/user-management/keycloak/), you are not using Dex.

The ability to support CLI logins is dependent on your OIDC provider. Typically you will need to configure a client in your OIDC provider to allow http://localhost:8080 as an OIDC client. Additionally, with CLI logins, a client secret may not be involved since they cannot be secured. Other providers such as Okta, require you to create a "native" application (as opposed to "web" application) and use PKCE (Proof Key for Code Exchange by OAuth Public Clients).

Perhaps someone on the Red Hat side (@jannfis) may know more about how to achieve CLI logins via Key Cloak.

**Reply by iam-veeramalla** [reply]
I will bring a keycloak setup with Argo CD in the morning and see what is going wrong :)

**Reply by gnometsunami** [reply]
> The ability to support CLI logins is dependent on your OIDC provider. Typically you will need to configure a client in your OIDC provider to allow http://localhost:8080 as an OIDC client. Additionally, with CLI logins, a client secret may not be involved since they cannot be secured. Other providers such as Okta, require you to create a "native" application (as opposed to "web" application) and use PKCE (Proof Key for Code Exchange by OAuth Public Clients).

CLI logins seem to use http://localhost:8085/auth/callback as the redirect url. I have added it to the valid redirects in keycloak. Maybe I can create a second client for the CLI in keycloak so that I can use different security configurations?

**Reply by iam-veeramalla** [reply]
Hi @ezdan , I was able to fix the issue by making the client public as shown below. 
<img width="1440" alt="Screenshot 2021-08-20 at 3 16 22 PM" src="https://user-images.githubusercontent.com/43399466/130215046-e00e2a7c-ca2f-47f2-82f6-e3295049b090.png">

If your security policies does not allow you to make the client public then yes, I think you would need to create a new client for CLI. Does this work for you ? I will be awaiting your reply.

**Reply by iam-veeramalla** [reply]
However, I see a documentation gap w.r.t the below 2 points.
1. Add the `http://localhost:8085/auth/callback` to the valid redirect URLS of the Keycloak Argo CD client.
2. Since the CLI access cannot be secured with the secret, we may need to make the client public or create a new public client for the CLI access.

I will update the documentation if @ezdan confirms that the above changes works :)

**Reply by gnometsunami** [reply]
Yes! Changing the client to public enabled me to sign in with cli. I'm not sure what the security implications are for the web ui, but at least the cli is functional.

**Reply by jessesuen** [reply]
I see that there is an "Implicit Flow" option in KeyCloak. As an alternative to public access type, can either of you try enabling Implicit flow to see if it works (in addition to whitelisting http://localhost:8085/auth/callback)?

> I'm not sure what the security implications are for the web ui, but at least the cli is functional.

The implications are: if an attacker was successfully able to perform a MITM (man in the middle) attack and masquerade as your Argo CD, they would then be allowed to complete OAuth2 login flows as your users login, because your KeyCloak will not care that the hijacked Argo CD is not supplying a valid client secret.

**Reply by gnometsunami** [reply]
Implicit doesnt seem to help.

access: confidential
implicit flow: enabled
```
FATA[0002] oauth2: cannot fetch token: 401 Unauthorized
Response: {"error":"unauthorized_client","error_description":"Client secret not provided in request"}
```

I'll invoke the 80/20 rule and just leave it as public. I feel that the mitm risk is pretty low on my private network.

@iam-veeramalla if you copy your answer in a new comment, I'll mark it as answered.

**Comment by iam-veeramalla** [comment, marked as answer]
Hi @ezdan , I was able to fix the issue by making the client public as shown below.
<img width="1440" alt="Screenshot 2021-08-20 at 3 16 22 PM" src="https://user-images.githubusercontent.com/43399466/130400482-dfba7a1f-503f-45bd-83c0-c8bb1ec9c016.png">

If your security policies does not allow you to make the client public then yes, I think you would need to create a new client for CLI. Does this work for you ? I will be awaiting your reply.

**Reply by gnometsunami** [reply]
Thank you! And thanks @jessesuen !

**Reply by 862902354** [reply]
Hi!
The keycloak 20 version has no access type: public option. How do you solve it?

**Reply by richardbeumer** [reply]
In keycloak 20 and above you  can make your client `public` by setting `Client authentication` to off

**Comment by vipunsanjana** [comment]
yeah abhishek .. some times i cant to create it..can you give me solution
